src/rdb.c

```diff
@@ -2362,6 +2362,13 @@ int rdbLoadRio(rio *rdb, int rdbflags, rdbSaveInfo *rsi) {
         sds key;
         robj *val;
 
+        /* When a new master is specified during slave rdbLoad, we should stop load
+         * the old RDB and sync new Master ASAP. */
+        if ((rdbflags & RDBFLAGS_REPLICATION) && server.repl_state != REPL_STATE_TRANSFER) {
+          serverLog(LL_WARNING, "Repl state changed, Stop load the RDB");
```

Comment:
If we are loading a large object, we can process events multiple times. That would give the replication handshake enough time to get to REPL_STATE_TRANSFER state and not abort here. I think this would cover most cases, but it might be worth making a little more robust?